src/dbnode/persist/fs/seek_manager.go

```diff
@@ -269,11 +275,54 @@ func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentD
 	}
 
 	found := false
-	for i, compareSeeker := range seekersAndBloom.seekers {
+	for i, compareSeeker := range seekers.active.seekers {
+		if seeker == compareSeeker.seeker {
+			found = true
+			compareSeeker.isBorrowed = false
+			seekers.active.seekers[i] = compareSeeker
+			break
+		}
+	}
+
+	if found {
+		return nil
+	}
+
+	// If no match was found in the active seekers, its possible that an inactive seeker is being returned.
+	for i, compareSeeker := range seekers.inactive.seekers {
 		if seeker == compareSeeker.seeker {
 			found = true
 			compareSeeker.isBorrowed = false
-			seekersAndBloom.seekers[i] = compareSeeker
+			seekers.inactive.seekers[i] = compareSeeker
+
+			// The goroutine that returns the last outstanding inactive seeker is responsible for notifying any
+			// goroutines waiting for all inactive seekers to be returned and clearing out the inactive seekers
+			// state entirely.
+			allAreReturned := true
+			for _, inactiveSeeker := range seekers.inactive.seekers {
+				if inactiveSeeker.isBorrowed {
+					allAreReturned = false
+					break
+				}
+			}
+
+			if !allAreReturned {
+				break
+			}
+
+			// All the inactive seekers have been returned so its safe to signal and clear them out.
+			var multiErr = xerrors.NewMultiError()
+			for _, inactiveSeeker := range seekers.inactive.seekers {
+				multiErr = multiErr.Add(inactiveSeeker.seeker.Close())
+			}
+			// Clear out inactive state.
+			seekers.inactive = seekersAndBloom{}
+			// Signal completion regardless of any errors encountered while closing.
+			seekers.inactive.wg.Done()
+
+			if multiErr.FinalError() != nil {
+				return multiErr.FinalError()
```

Comment:
```go
if err := multiErr.FinalError(); err != nil {
    return err
}
```
But don't you always just want to `return multiErr.FinalError()` here instead of having an if statement at all since you break right after anyway?